### PR TITLE
[#1012] Introduce dependency injection

### PR DESCRIPTION
**Pull request by Chlup**
- This is the first change for the dependency injection. This change adds `DIContainer` which is used to register and resolve dependencies.
- Then this change shows how `DIContainer` can be used to resolve dependencies. And it also shows how to mock dependencies in tests.
- Constructors of `Initializer` are changed to support mocking of depedendencies. But it doesn't affect public API in any way.
- Depedencies are registered in `Dependencies.setup()`. Then in the code `container.resolve(SomeType.self)` is called to get instance of the dependency.
- `ZcashTestCase` class is added. It inherits from `XCTestCase` and should be used for base class in tests. `ZcashTestCase` takes care of creation of `DIContainer` for dependencies mocking. In future we can maybe move there more things that are used in each test.
- Lot is missing here. Not all the code is using dependency injection. Tests for `DIContainer` are missing. Only `OfflineTests` work now (other tests can't be even compiled). It will be added in future changes.

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [ ] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Utils/DIContainer.swift [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        lock.lock()
+        defer { lock.unlock() }
+        let key = self.key(for: type)
+
+        let possibleDependency = (isTestEnvironment ? mockedDependencies[key] : dependencies[key]) ?? dependencies[key]
+        guard let dependency = possibleDependency else {
+            // When dependency is resolved before it's registered then the app crashes. It would be possible to not crash and throw some error here.
+            // But it complicates the rest of the code. And maybe it doesn't make sense because this kind of error is not recoverable. It can be fixed
+            // only by updating the code.
+            fatalError("Doesn't have registered dependency for type \(type).")
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Utils/DIContainer.swift:64]: Please pay attention to this part and let me know what do you think.

File: Sources/ZcashLightClientKit/Utils/DIContainer.swift [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        if dependency.isSingleton && dependency.instance == nil {
+            dependencies[key] = Dependency(factory: dependency.factory, isSingleton: dependency.isSingleton, instance: instance)
+        }
+
+        guard let instance = instance as? T else {
+            // When dependency is resolved but instance of created depedency is different than expected type the app crashes. It would be possible to
+            // not crash and throw some error here. But it complicates the rest of the code. And maybe it doesn't make sense because this kind of
+            // error is not recoverable. It can be fixed only by updating the code.
+            fatalError("Getting dependency for type \(type) but created instance is \(instance)")
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Utils/DIContainer.swift:82]: Please pay attention to this part and let me know what do you think.

File: Sources/ZcashLightClientKit/Utils/DIContainer.swift (hunk 1 above)
Comment (pacu) [reply Sources/ZcashLightClientKit/Utils/DIContainer.swift:64]: This is a programming error so it's correct to crash a foreseeable unstable application.

File: Sources/ZcashLightClientKit/Utils/DIContainer.swift (hunk 2 above)
Comment (pacu) [reply Sources/ZcashLightClientKit/Utils/DIContainer.swift:82]: Same as above. This catches programming errors or purposely introduced inconsistencies. It is correct to crash the application on clear programming error.

**Review by pacu — APPROVED** [review]

**Pull request merged**